consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/CliqueBlockHeaderFunctions.java

```diff
@@ -30,4 +30,9 @@ public Hash hash(final BlockHeader header) {
   public CliqueExtraData parseExtraData(final BlockHeader header) {
     return CliqueExtraData.decodeRaw(header);
   }
+
+  @Override
+  public int getCheckPointWindowSize(final BlockHeader blockHeader) {
+    return CliqueExtraData.decode(blockHeader).getValidators().size();
```

Comment:
this seems maximally safe but why is it necessary if we are specifying block number, hash and difficulty?  I am guessing it is PoA related just not sure why.

Reply:
this is clearly a POA specificity. If you look at this code. you will see that to validate a signer. We need to look at n block before to check that a validator has not already signed a block (where n is the number of validators at the moment of this block). This is not reset during a new epoch. So the easiest I found is to download the N blocks before in order to have the necessary information for the checkpoint and the following blocks

https://github.com/hyperledger/besu/blob/ed1329cf848a9d7858d039a55add810da3c428b5/consensus/clique/src/main/java/org/hyperledger/besu/consensus/clique/headervalidationrules/SignerRateLimitValidationRule.java#L35